Support pointer data types for sizeof operator

ChAoSUnItY opened the pull request:
Currently `sizeof` is only capable of parsing basic data types like `int`, `char`, or user-defined data types like `struct`, `enum` etc. But not able to parsing pointer type, thus, in this PR, by introducing a temporary workaround parsing 0 or more asterisk tokens, we now partially support pointer type for `sizeof` operator (not including function pointer type yet).

Additionally, since the result given by `HOST_PTR_SIZE` in stage 0 and stage 1/2 are different, so I make another test driver named `driver-stage2.sh` for testing stage 2 shecc and adding corresponding `sizeof` test to verify the implementation.

jserv commented on Makefile:
```diff
@@ -52,12 +52,20 @@ $(OUT)/tests/%.elf: tests/%.c $(OUT)/$(STAGE0)
+	$(VECHO) "  TEST SNAPSHOTS\n"
 	tests/check-snapshots.sh
 
+check-all: $(OUT)/$(STAGE2) $(TESTBINS) check check-stage2
```
Don't name it `check-all`. Instead, consolidate the existing `check` for stage1 and stage2.

jserv commented on tests/driver-stage2.sh:
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env bash
```
I have no idea why `tests/driver-stage2.sh` exits.

jserv replied:
Is it possible to stick to single `tests/driver.sh` for validation?

ChAoSUnItY replied:
Yes, it would be possible. Just need to plug in 1 one argument to specify which stage's executable is being used with testing.

ChAoSUnItY replied:
I have tested `tests/driver.sh` with stage 2, but I have found several problems:
1. The compilation speed is significantly slower than stage 0.
2. Stage 2 compiler would failed to compile these test suites
https://github.com/sysprog21/shecc/blob/18e005dbe21b323dc1559195388e02e7ff101bba/tests/driver.sh#L955-L985

DrXiao replied:
> Stage 2 compiler would failed to compile these test suites

I use the following test code to test shecc first.
```c
int main() 
{ 
    printf("%d %d\n", -2147483648 / 1, -2147483648 % 1); 
    return 0;  
} 
```
Using current shecc to compile the code would result in a Bus Error, and I believe the potential cause is `__stmdb` and `__ldm` in the Arm backend, which operate on memory to store or load multiple words, .

If temporarily removing these instructions from `arm-codegen.c`, shecc will compile the test code successfully.
```diff
    case OP_div:
    case OP_mod:
        ...
        interm = __r8;
        /* div/mod emulation */
        /* Preserve the values of the dividend and divisor */
-       emit(__stmdb(__AL, 1, __sp, (1 << rn) | (1 << rm)));
        /* Obtain absolute values of the dividend and divisor */
        emit(__srl_amt(__AL, 0, arith_rs, __r8, rn, 31));
        ...
        emit(__mov_r(__AL, __r9, rn));
-       emit(__ldm(__AL, 1, __sp, (1 << rn) | (1 << rm)));
        emit(__mov_r(__AL, rd, interm));
        ...
        return;
```
---

That is, the behavior of `__stmdb` and `__ldm` may lead to unaligned memory access, thereby triggering a `Bus Error`.

Therefore, I think improving the current Arm bakcend could be a solution to resolve the issue.

reference: [ARM Cortex-A Series Programmer's Guide for ARMv7-A](https://developer.arm.com/documentation/den0013/d/Porting/Alignment)
> However, load and store multiple instructions (LDM and STM) and load and store double-word (LDRD or STRD) must be aligned to at least a word boundary.

ChAoSUnItY replied:
Currently I have resolved ARM stack misalignment by ensure `__stmdb` and `__ldm` pushes 4 registers at a time to keep it aligned.

But now the RISCV backend will stuck, but I found that current implementation does not check the formerly mentioned test suite edge case scenario: `INT_MIN` could not be approximated due to missing equality check, which leads the program keep shifting the approximated value after reaching `INT_MIN` then causes the approximated value always be 0 afterwards.

By adding additional equality test while approximating nearest power of 2 value, the program would work as expected.

DrXiao commented on tests/driver-stage2.sh:
```diff
@@ -0,0 +1,110 @@
+expr 4 "sizeof(int)";
+expr 1 "sizeof(char)";
+# sizeof pointers
+expr 4 "sizeof(void*)";
```
According to your proposed changes, I think here can also adds the following cases:
```c
/* size of a pointer to pointer to integer or other basic data types */
sizeof(int **);
```
```c
typedef struct {
    int a;
    int b;
} test_struct;

 /* size of a pointer to a structure */
sizeof(test_struct *);
```

ChAoSUnItY commented:
I'm unable to reproduce broken pipe failure on my machine, and the snapshots I checked which are generated by `tests/update-snapshots.sh` are same as the original one.

DrXiao commented:
Because this pull request involves size of pointers, I would like to ask a question.

In the header file (`def.h`), we can notice that there are two macros called `PTR_SIZE`([Line 47](https://github.com/sysprog21/shecc/blob/18e005dbe21b323dc1559195388e02e7ff101bba/src/defs.h#L47)) and `HOST_PTR_SIZE`  ([Line 59 ~ 64](https://github.com/sysprog21/shecc/blob/18e005dbe21b323dc1559195388e02e7ff101bba/src/defs.h#L59)). Obviously, both of them mean the size of a pointer, but I cannot understand their use and difference all the time, especially the later one (`HOST_PTR_SIZE`).

Thus, I hope to obtain an explanation about them through this opportunity.

ChAoSUnItY commented:
> In the header file (def.h), we can notice that there are two macros called PTR_SIZE ([Line 47](https://github.com/sysprog21/shecc/blob/18e005dbe21b323dc1559195388e02e7ff101bba/src/defs.h#L47)) and HOST_PTR_SIZE ([Line 59 ~ 64](https://github.com/sysprog21/shecc/blob/18e005dbe21b323dc1559195388e02e7ff101bba/src/defs.h#L59)). Obviously, both of them mean the size of a pointer, but I cannot understand their use and difference all the time, especially the later one (HOST_PTR_SIZE).

The `HOST_PTR_SIZE` is determined by the running host, while `PTR_SIZE` is supposed to be 4, which is capable of handling the arch we're targeting to. The only usages of `HOST_PTR_SIZE` currently are `memcpy` in `src/ssa.c`.

Which I later realized value of `sizeof(type*)` should not be `HOST_PTR_SIZE`, instead, it should be `PTR_SIZE`.

DrXiao commented:
@ChAoSUnItY , after fetching your latest changes, I try to build shecc and check the snapshots:
```shell
$ make distclean && make ARCH=arm && tests/check-snapshots.sh 
...
  SHECC	out/shecc-stage1.elf
  SHECC	out/shecc-stage2.elf
```
```shell
$ make distclean && make ARCH=riscv && tests/check-snapshots.sh 
...
  SHECC	out/shecc-stage1.elf
  SHECC	out/shecc-stage2.elf
Files /dev/fd/63 and /dev/fd/62 differ
Files /dev/fd/63 and /dev/fd/62 differ
```
It seems that the snapshots differs when shecc uses RISC-V backend. Hope this information helps you.

ChAoSUnItY commented:
Ok, I found the cause, apparently previously shecc's CI won't check snapshot on rv32 but only on arm32:

https://github.com/sysprog21/shecc/blob/18e005dbe21b323dc1559195388e02e7ff101bba/Makefile#L55-L59

https://github.com/sysprog21/shecc/blob/18e005dbe21b323dc1559195388e02e7ff101bba/.github/workflows/main.yml#L6-L27

So currently the workaround is just not to run check-snapshot on rv32.

Should we later address this issue in anther PR? @jserv

jserv commented:
> Ok, I found the cause, apparently previously shecc's CI won't check snapshot on rv32 but only on arm32:

That was a fault. Let's fix it in another pull request.

ChAoSUnItY commented:
I will wrap up some additional changes introduced besides `sizeof` operator implementation enhancement:

Previously, while using stage 2 compiler to compile test suites in `tests/driver.sh`, the arm32 suffers bus error, and the rv32 suffers infinite loop.

The former one was caused by improper stack alignment, while the later one was caused by missing edge case equality check, which is `INT_MIN`. Thus, I have added `stack_align` function in `reg_alloc.c` to ensure every stack increment is properly aligned, and I have added a branching instruction for the edge case.

The change of total stack size of shecc itself has increased from 9225 to 9228.

DrXiao commented on src/arm-codegen.c:
```diff
@@ -336,7 +336,9 @@ void emit_ph2_ir(ph2_ir_t *ph2_ir)
         interm = __r8;
         /* div/mod emulation */
-        /* Preserve the values of the dividend and divisor */
+        /* Preserve the values of the dividend and divisor. This also keeps
```
Because the latest changes preserve the original use of `__stmdb` and `__ldm`, I think the comment doesn't need to be changed.

jserv commented:
I defer to @vacantron for confirmation.

vacantron reviewed (approved):
LGTM

jserv commented:
Thank @ChAoSUnItY for contributing!
